File: swift/plugins/Secretspec.swift
// The `secretspec` provider kind: SecretSpec, through its CLI.
//
// A plugin, not a built-in: it spawns a child process.
//
// A port of typescript/plugins/secretspec.ts, which is canonical.

import Foundation

import Sekreto

// A SCOPED import: plugin exports a `Json` of its own, and the only name
// this file wants from it is `Definition`.
import struct VoxgigPlugin.Definition

/// SecretSpec (https://secretspec.dev).
///
/// SecretSpec is a declaration - a `secretspec.toml` naming the secrets a
/// project needs - plus a chain of its own backends to satisfy them from.
/// That makes it the same shape as sekreto one level down, and the reason
/// to support it is the same reason sekreto exists: a project that has
/// already declared its secrets there should not have to declare them
/// again here.
///
/// `backend` selects one of SecretSpec's backends (`--provider`) and is
/// called `backend` here only because `provider` already means something
/// else in this library.
///
/// A reason is required, not optional: SecretSpec records every read in an
/// audit log and refuses to read at all without one.
public final class SecretspecProvider: Provider {

  private let command: String
  private let file: String?
  private let profile: String?
  private let backend: String?
  private let reason: String?
  private let prefix: String?

  public init(
    command: String? = nil,
    file: String? = nil,
    profile: String? = nil,
    backend: String? = nil,
    reason: String? = nil,
    prefix: String? = nil
  ) {
    self.command = first(command, "secretspec")
    self.file = file
    self.profile = profile
    self.backend = backend
    self.reason = reason
    self.prefix = prefix
  }

  public func lookup(_ name: String) throws -> String? {
    let key = try envkey(name, prefix)

    // `--file` comes BEFORE the subcommand; everything else after it.
    var argv = [command]
    if let use = file, !use.isEmpty { argv += ["--file", use] }
    argv += ["get", key]
    if let use = backend, !use.isEmpty { argv += ["--provider", use] }
    if let use = profile, !use.isEmpty { argv += ["--profile", use] }
    argv += ["--reason", first(reason, "sekreto")]

    let ran = try runcmd(argv, ProcessInfo.processInfo.environment, command)

    if 0 == ran.status {
      // The value and one newline, and nothing else.
      return dropsuffix(ran.out, "\n")
    }

    if secretspecmiss(ran.why, key) { return nil }

    throw SekretoError(
      "sekreto: secretspec error: " + (ran.why.isEmpty ? "exit \(ran.status)" : ran.why))
  }

  public func describe() -> String {
    let use = backend ?? ""
    return use.isEmpty ? "secretspec" : "secretspec:\(use)"
  }
}

/// Does this SecretSpec failure mean "no such secret" rather than "I could
/// not answer"?
///
/// MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
/// `Provider backend 'keyring' not found`, which is a store that could not
/// answer at all - and reading that as a miss is the worst failure this
/// library has, because the chain then falls through to a weaker store
/// without saying so. The key is required to appear, so the two cannot be
/// confused.
public func secretspecmiss(_ why: String, _ key: String) -> Bool {
  return why.contains("Secret '\(key)' not found")
}

/// The kind, as a voxgig/plugin definition.
public let secretspec: Definition = providerplugin("secretspec") { spec in
  SecretspecProvider(
    command: spec.command,
    file: spec.file,
    profile: spec.profile,
    backend: spec.backend,
    reason: spec.reason,
    prefix: spec.prefix
  )
}

File: dart/plugins/secretspec.dart
// SecretSpec, as a voxgig/plugin definition.
//
// A PLUGIN, not a built-in: it spawns a child process. A `Sekreto` can
// build the `secretspec` kind only if the calling project imported this
// file and passed `secretspec` in the `plugins` option
// (docs/design/plugin-providers.md).
//
// A port of typescript/plugins/secretspec.ts, which is canonical.

import '../src/provider.dart';
import '../src/sekreto.dart';
import '../src/support.dart';

import 'httpjson.dart';

/// Does this SecretSpec failure mean "no such secret" rather than "I could
/// not answer"?
///
/// SecretSpec says `Secret 'API_TOKEN' not found` for both a name it does
/// not declare and one declared with no value, and both are misses: this
/// store does not hold it, so the chain carries on.
///
/// MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
/// `Provider backend 'keyring' not found`, which is a store that could not
/// answer at all - and reading that as a miss is the worst failure this
/// library has, because the chain then falls through to a weaker store
/// without saying so. The key is required to appear, so the two cannot be
/// confused.
bool secretspecmiss(String reason, String key) =>
    reason.contains("Secret '$key' not found");

/// SecretSpec (https://secretspec.dev).
///
/// SecretSpec is a declaration - a `secretspec.toml` naming the secrets a
/// project needs - plus a chain of its own backends to satisfy them from.
/// That makes it the same shape as sekreto one level down, and the reason to
/// support it is the same reason sekreto exists: a project that has already
/// declared its secrets there should not have to declare them again here.
///
/// A reason is required, not optional: SecretSpec records every read in an
/// audit log and refuses to read at all without one. sekreto sends `sekreto`
/// unless told otherwise, so the audit trail says which tool asked.
class Secretspec extends Provider {
  final String command;
  final String? file;
  final String? profile;
  final String? backend;
  final String? reason;
  final String? prefix;

  Secretspec({
    String? command,
    this.file,
    this.profile,
    this.backend,
    this.reason,
    this.prefix,
  }) : command = (null == command || command.isEmpty) ? 'secretspec' : command;

  @override
  String? lookup(String name) {
    final key = envkey(name, prefix);

    // The order is SecretSpec's: --file belongs to the program, before the
    // subcommand; everything else belongs to `get`.
    final argv = <String>[command];
    if (null != file && file!.isNotEmpty) {
      argv.addAll(['--file', file!]);
    }
    argv.addAll(['get', key]);
    if (null != backend && backend!.isNotEmpty) {
      argv.addAll(['--provider', backend!]);
    }
    if (null != profile && profile!.isNotEmpty) {
      argv.addAll(['--profile', profile!]);
    }
    argv.addAll(['--reason', first([reason, 'sekreto'])]);

    final ran = runcmd(argv, command);

    if (0 == ran.status) {
      // The value and one newline, and nothing else.
      return dropsuffix(ran.out, '\n');
    }

    if (secretspecmiss(ran.why, key)) {
      return null;
    }

    throw SekretoError(
      'sekreto: secretspec error: '
      '${ran.why.isEmpty ? 'exit ${ran.status}' : ran.why}',
    );
  }

  @override
  String describe() =>
      'secretspec${(null != backend && backend!.isNotEmpty) ? ':$backend' : ''}';
}

/// The `secretspec` provider kind. Pass it in `plugins` to name it in a
/// chain.
final Definition secretspec = providerplugin(
  'secretspec',
  (spec) => Secretspec(
    command: spec.command,
    file: spec.file,
    profile: spec.profile,
    backend: spec.backend,
    reason: spec.reason,
    prefix: spec.prefix,
  ),
);
